packages/eslint-plugin/src/rules/no-void-expression.ts

```diff
@@ -0,0 +1,284 @@
+        if (invalidAncestor.type === AST_NODE_TYPES.ArrowFunctionExpression) {
+          // handle wrapping with `void`
+          if (options.ignoreVoidOperator) {
+            return context.report({
+              node,
+              messageId: 'invalidVoidExprArrowWrapVoid',
+              fix: wrapVoidFix,
+            });
+          }
+
+          // handle wrapping with braces
+          const arrowFunction = invalidAncestor;
+          return context.report({
+            node,
+            messageId: 'invalidVoidExprArrow',
+            fix(fixer) {
+              const arrowBody = arrowFunction.body;
+              const arrowBodyText = sourceCode.getText(arrowBody);
+              const newArrowBodyText = `{ ${arrowBodyText}; }`;
+              if (isParenthesized(arrowBody, sourceCode)) {
+                const [bodyOpeningParen, bodyClosingParen] = [
+                  sourceCode.getTokenBefore(arrowBody, isOpeningParenToken)!,
+                  sourceCode.getTokenAfter(arrowBody, isClosingParenToken)!,
+                ];
+                return fixer.replaceTextRange(
+                  [bodyOpeningParen.range[0], bodyClosingParen.range[1]],
+                  newArrowBodyText,
+                );
+              }
+              return fixer.replaceText(arrowBody, newArrowBodyText);
+            },
+          });
+        }
+
+        // handle return statement
+        if (invalidAncestor.type === AST_NODE_TYPES.ReturnStatement) {
+          // handle wrapping with `void`
+          if (options.ignoreVoidOperator) {
+            return context.report({
+              node,
+              messageId: 'invalidVoidExprReturnWrapVoid',
+              fix: wrapVoidFix,
+            });
+          }
+
+          // handle moving to it's own statement
+          const returnStmt = invalidAncestor;
+          let isTopLevelReturn = false;
+          let isLastReturn = false;
+          if (returnStmt.parent?.type === AST_NODE_TYPES.BlockStatement) {
+            const block = returnStmt.parent;
+            const blockType = block.parent?.type;
+            if (
+              blockType === AST_NODE_TYPES.FunctionDeclaration ||
+              blockType === AST_NODE_TYPES.FunctionExpression ||
+              blockType === AST_NODE_TYPES.ArrowFunctionExpression
+            ) {
+              isTopLevelReturn = true;
+            }
+            if (block.body.indexOf(returnStmt) === block.body.length - 1) {
+              isLastReturn = true;
+            }
+          }
+
+          // remove the `return` keyword
+          if (isTopLevelReturn && isLastReturn) {
+            return context.report({
+              node,
+              messageId: 'invalidVoidExprReturnLast',
+              fix(fixer) {
+                const returnValue = returnStmt.argument!;
+                const returnValueText = sourceCode.getText(returnValue);
+                const newReturnStmtText = `${returnValueText};`;
```

Comment:
I have a concern about this possibly tripping people who don't use semi colons, for instance this:

```ts
function foo(){
   const x = "hello"
   return [1,2,3].forEach(v=>console.log(`${x} ${v}`)
}
```
when the return is removed this is changed to semantically different (and invalid) code.  While this kind of case is exactly why semi colons are really good idea, there are people who may be harmed by this and I don't think there is really anything we could or should do about it. Just wanted to bring it up.

Reply:
Good catch. Even I prefer no semicolons and didn't notice that.

TODO: I will work on this later.

Reply:
Fixed. It now adds a semicolon at the beginning of the line if the expression starts with `(`, `[` or `` ` ``.